crates/environ/src/module.rs

```diff
@@ -259,6 +259,100 @@ impl ModuleTranslation<'_> {
+            // we have.
+            if table_index.index() < num_imported_tables {
+                continue;
+            }
+
+            // If this is not a funcref table, then we can't support a
+            // pre-computed table of function indices.
+            if self.module.table_plans[table_index].table.wasm_ty != WasmType::FuncRef {
+                continue;
+            }
+
+            // If there is no initial content for the table, then
+            // there is nothing to do here.
+            let init = match init {
+                Some(init) => init,
+                None => continue,
+            };
+
+            // Get the list of segment initializers. If we're in some
+            // other mode, then nothing to do here.
+            let segments = match init.as_segments() {
+                Some(segments) => segments,
+                None => continue,
+            };
+
+            // Check all segments to see if any have a dynamic base
+            // index. If any are, then we can't build an array of
+            // initial contents that is independent of instantiation
+            // state (values of globals), so remain in Segments mode.
+            let has_dynamically_placed_segment =
+                segments.iter().any(|segment| segment.base.is_some());
+            if has_dynamically_placed_segment {
+                continue;
+            }
+
+            // Check the table size. If too large to justify a dense
+            // data structure (array with entries for all elements),
+            // then remain in Segments mode.
+            let table_size = self.module.table_plans[table_index].table.minimum;
+            if table_size > MAX_FUNC_TABLE_SIZE {
+                continue;
+            }
+
+            // Check all segments to see if any are out-of-bounds. If
+            // any are, then we fall back to the default code-paths to
+            // handle the error and get the final state just before
+            // the trap correct.
+            let has_out_of_bounds = segments.iter().any(|segment| {
+                segment
+                    .offset
+                    .checked_add(segment.elements.len() as u32)
+                    .unwrap_or(u32::MAX)
+                    > table_size
+            });
+            if has_out_of_bounds {
+                continue;
+            }
+
+            // Now we can commit to building an array of FuncIndex
+            // values.
+            let mut elements: Vec<Option<FuncIndex>> = vec![None; table_size as usize];
+
+            for segment in segments {
+                let base = segment.offset as usize;
+                // Can't overflow: already checked above.
+                let top = base + segment.elements.len();
+
+                let dst = &mut elements[base..top];
+                for (dst, src) in dst.iter_mut().zip(segment.elements.iter()) {
+                    *dst = Some(*src);
```

Comment:
Actually now that I think about this in relation to my above comment I don't think that this is quite right in that element segments can have `ref.null` as a particular element which is represented with `FuncIndex::reserved_value()` which gets set to `Some` here as well. That's probably handled gracefully elsewhere but in any case I think it'd be good to avoid the `Option<FuncIndex>` here

Reply:
Done!